New Routing Documentation (#424)
CloudCannon/platform-documentation: articles/configure-custom-routing.mdx

---
_schema: default
_uuid: 9c25997c-8237-4f48-a474-037304cd3982
_created_at: '2021-09-15 23:13:40 +0000'
title: Configure custom routing
nav_title: Configure custom routing
published: true
image: >-
  https://cc-dam.imgix.net/documentation/images/default-thumbnail/2023-Q4/CloudCannon-Documentation-Thumbnail.png
description: Configure your CloudCannon's hosting by configuring a single json file.
article_category:
  - Instructions
article_topic:
  - Hosting
  - Routing
tags:
  - Hosting
  - Sites
  - Redirects
  - Custom Headers
  - Guides
related_articles: []
related_links:
---
<comp.Notice info_type="important">
  Custom routing is only available for Sites [hosted through CloudCannon](/documentation/articles/what-is-web-hosting/#hosting-on-cloud-cannon). If you host your Site externally, or use CloudCannon in [Headless Mode](/documentation/articles/what-is-headless-mode/), custom routing will not work.
</comp.Notice>

> Custom routing allows you to control which content is delivered to your visitors when they enter a URL in their internet browser. You can create custom routing by creating a `routing.json` file in your `.cloudcannon` folder. When a visitor requests your webpage, the file server will also load the `routing.json` file to extend the default routing behavior.

Reviewer: "the file server" might be too abstract here without context — replace with "CloudCannon's server" or just "CloudCannon" ?

> Your custom routing file can specify [redirects](/documentation/articles/configure-custom-routing/#custom-redirects) and [headers](/documentation/articles/configure-custom-routing/#custom-headers) using JSON code.

Reviewer: ```suggestion
Your custom routing file can specify [redirects](/documentation/articles/configure-custom-routing/#custom-redirects) and [headers](/documentation/articles/configure-custom-routing/#custom-headers) using JSON data.
```

* **Custom redirect** — an instruction to redirect a visitor from one URL to another.
* **Custom header** — additional information passed between the web server and the client (e.g., visitor's browser).

<comp.Notice info_type="info">

>   Invalid JSON code in the routing.json file will cause your Site [builds](/documentation/articles/what-is-a-site-build/) to fail, protecting the live version of your Site. You can troubleshoot your JSON using the [build output logs](/documentation/articles/what-is-a-site-build/#reading-your-build-output-logs).

Reviewer: ```suggestion
  Invalid JSON data in the routing.json file will cause your Site [builds](/documentation/articles/what-is-a-site-build/) to fail, protecting the live version of your Site. You can troubleshoot your JSON using the [build output logs](/documentation/articles/what-is-a-site-build/#reading-your-build-output-logs).
```

* **From** — Specifies which URL should trigger this redirect rule.
* **To** — Specifies which URL the visitor should be redirected to.
* **Status** — Tells the visitor's browser the reason for the redirect.

CloudCannon allows the following [HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes):

* **200** — Indicates the request was successful.
* **301** — Indicates the requested URL has been changed permanently and specifies the new URL in the response.
* **302** — Indicates the requested URL has been changed temporarily, and the file server should continue to use the original URL in future requests.
* **307** — Indicates the requested URL has been changed temporarily and the file server should continue to use the original URL in future requests. This is similar to a 302 code however it does not allow the HTTP method to change.
* **308** — Indicates the requested URL has been changed permanently and the file server should continue to use the original URL in future requests. This is similar to a 301 code however it does not allow the HTTP method to change.
* **404** — Indicates the requested URL could not be found.

> * **410** — Indicates the requested URL has been permanently deleted, and search engines should de-index this page.

Reviewer: It might be worth making a distinction between `200` / `404` / `410` and the rest of them. These three act as proxies and don't redirect the user, instead they change the content that is served for the requested URL. The others are true redirects, and send the user to a different URL that they will see.

Reviewer: To expand on that — the `## Custom Redirects` should probably be `## Custom Routes`, with the `301`, `302`, `307`, `308` described as "redirects"

> Most users will only need to specify redirects using the 301, 302, and 404 status codes. We recommend redirecting your old page URLs with 301 or 302 redirects whenever you change the URL of a page or the architecture of your Site in a way that affects the URL structure. This helps to preserve your SEO scores if you want to rank highly in search engine results, maintains external links to your website, and prevents visitors from seeing incorrect or missing content when they attempt to load any of your old URLs.

Reviewer: ```suggestion
Most users will only need to specify redirects using the 307, 307, and 404 status codes. We recommend redirecting your old page URLs with 307 or 308 redirects whenever you change the URL of a page or the architecture of your Site in a way that affects the URL structure. This helps to preserve your SEO scores if you want to rank highly in search engine results, maintains external links to your website, and prevents visitors from seeing incorrect or missing content when they attempt to load any of your old URLs.
```

307 and 308 are the new "ideal" redirect codes. It doesn't matter much for sites on CloudCannon but best practice is to use them over 301 or 302.

> We also recommend [configuring a 404 page](/documentation/articles/configure-a-404-page/) and redirecting all unrecognized subpaths to that page.

Reviewer: ```suggestion
We also recommend [configuring a 404 page](/documentation/articles/configure-a-404-page/) and routing all unrecognized subpaths to that page.
```

One option for verbiage to use instead of redirect for the status codes that aren't "redirects"

<comp.Notice info_type="info">
  A custom redirect will only occur if the `from` URL does not resolve when a visitor attempts to load the page. CloudCannon reads your redirects in order, and will use the earliest matching rule for a given `from` URL. Duplicate rules are ignored.

> </comp.Notice>

Reviewer: * specify with `forced: false`

Reviewer 2: Changed the copy

<comp.Notice info_type="info">
  You can redirect to, but not from a URL fragment (e.g. example.com/blog/#heading). The URL fragment is not passed to the server so, it cannot be processed by the redirect.
</comp.Notice>

> <comp.Notice info_type="info">
>   You can use redirects to load a single HTML file on multiple URL routes and use JavaScript to load different content.
> </comp.Notice>

Reviewer: ```suggestion
<comp.Notice info_type="info">
  You can use routing to load a single HTML file on multiple URL routes and use JavaScript to load different content.
</comp.Notice>
```

3. When you are happy with your routing file, save your changes.

After a successful build, CloudCannon will redirect visitors according to the rules in your `routing.json` file.

Here are a few examples:

<comp.CodeBlock language="json" source=".cloudcannon/routing.json">
```

{
  "routes": [
    {
      "from": "/cloudcannon/(.*)",
      "to": "https://cloudcannon.com/$1",
      "status": 302
    },
    {
      "from": "/documentation/articles/choosing-your-ssg/",
      "to": "/documentation/articles/select-your-ssg/",
      "status": 301
    },
    {
      "from": "/app/(.*)",
      "to": "/app/",
      "status": 200
    }
  ]
}

```
</comp.CodeBlock>

### Generating redirects

If your Site requires it, you can generate your routing file at build time.

For example, if you only want to redirect URLs on your production Site, you could add the following Liquid templating:

<comp.CodeBlock language="JSON" source="routing.json">
``````````

{
  routes: [
    {% if environment == 'production' %}
      {from: '/example-1',
      to: '/example-2'}
    {% endif %}
  ]
}

``````````
</comp.CodeBlock>

> You can create a build process generate your routing file with the correct content, depending on the environment. In this case, the build process should generate the file in  `_cloudcannon/routing.json` . CloudCannon will use this file instead of  `.cloudcannon/routing.json` from your source.

Reviewer 3: ```suggestion
You can create a build process to generate your routing file with the correct content, depending on the environment. In this case, the build process should generate the file in  `_cloudcannon/routing.json` . CloudCannon will use this file instead of  `.cloudcannon/routing.json` from your source.
```

For more information, please contact our friendly [support team](/support/).

## Custom Headers

Custom HTTP headers allow you to pass more information between your web server and the visitor's browser every time someone loads a web page. Custom headers are often used to change some website behavior, meet security requirements, or allow other origins to load resources.

In CloudCannon, you can configure headers using the `routing.json` file. Whenever a visitor loads a web page from your Site, all relevant header rules from your routing file are applied in order.

To add custom headers to your `routing.json` file:

1. Navigate to the `.cloudcannon` folder in your *File browser* and click on the `routing.json` file. CloudCannon will open `routing.json` in the Source Editor.
2. Create a JSON array under the `headers` key.

> 3. Within the `headers` key, add an object labeled `match` and set it to `*`, and a second `headers` array.

Reviewer: Again specify that this `match` is an RE2 regex — this example is specifically adding custom headers to _every_ URL, but you could also specify custom headers for specific URLs.

4. Within the second `headers` array, add an entry for each custom header. Each entry in the array should include the keys `name` and `value`.
5. When you are happy with your routing file, save your changes.

After a successful build, CloudCannon will load your custom headers whenever a page loads.

<comp.CodeBlock language="json" source=".cloudcannon/routing.json">
```

{
  "headers": [
    {
      "match": "*",
      "headers": [
        {
          "name": "Access-Control-Allow-Origin",
          "value": "https://example.com"
        },
        {
          "name": "X-Content-Type-Options",
          "value": "nosniff"
        },
        {
          "name": "X-Robots-Tag",
          "value": "noindex, nofollow"
        }
      ]
    }
  ]
}

```

> </comp.CodeBlock>

Reviewer: We should document the allowed headers here — we don't allow _anything_ to be set. Currently supported:

- strict-transport-security
- x-robots-tag
- x-content-type-options
- x-frame-options
- x-xss-protection
- access-control-allow-origin
- access-control-allow-headers
- access-control-allow-methods
- content-security-policy
- content-security-policy-report-only
- expect-ct
- sourcemap
- large-allocation
- content-type
- referrer-policy
- permissions-policy

Reviewer 2: Added a notice. The full list will be available on the reference page.

---
_schema: default
_uuid: 9c25997c-8237-4f48-a474-037304cd3982
_created_at: '2021-09-15 23:13:40 +0000'
title: Configure custom routing
nav_title: Configure custom routing
published: true
image: >-
  https://cc-dam.imgix.net/documentation/images/default-thumbnail/2023-Q4/CloudCannon-Documentation-Thumbnail.png
description: Configure your CloudCannon's hosting by configuring a single json file.
article_category:
  - Instructions
article_topic:
  - Hosting
  - Routing
tags:
  - Hosting
  - Sites
  - Redirects
  - Custom Headers
  - Guides
related_articles: []
related_links:
---
<comp.Notice info_type="important">
  Custom routing is only available for Sites [hosted through CloudCannon](/documentation/articles/what-is-web-hosting/#hosting-on-cloud-cannon). If you host your Site externally, or use CloudCannon in [Headless Mode](/documentation/articles/what-is-headless-mode/), custom routing will not work.
</comp.Notice>

> Custom routing allows you to control which content is delivered to your visitors when they enter a URL in their internet browser. You can create custom routing by creating a `routing.json` file in your `.cloudcannon` folder. When a visitor requests your webpage, the file server will also load the `routing.json` file to extend the default routing behavior.

Reviewer 2: ```suggestion
Custom routing allows you to control which content is delivered to your visitors when they enter a URL in their internet browser. You can create custom routing by creating a `routing.json` file in your `.cloudcannon` folder. When a visitor requests your webpage, the CloudCannon will also load the `routing.json` file to extend the default routing behavior.
```

---
_schema: default
_uuid: 9c25997c-8237-4f48-a474-037304cd3982
_created_at: '2021-09-15 23:13:40 +0000'
title: Configure custom routing
nav_title: Configure custom routing
published: true
image: >-
  https://cc-dam.imgix.net/documentation/images/default-thumbnail/2023-Q4/CloudCannon-Documentation-Thumbnail.png
description: Configure your CloudCannon's hosting by configuring a single json file.
article_category:
  - Instructions
article_topic:
  - Hosting
  - Routing
tags:
  - Hosting
  - Sites
  - Redirects
  - Custom Headers
  - Guides
related_articles: []
related_links:
---
<comp.Notice info_type="important">
  Custom routing is only available for Sites [hosted through CloudCannon](/documentation/articles/what-is-web-hosting/#hosting-on-cloud-cannon). If you host your Site externally, or use CloudCannon in [Headless Mode](/documentation/articles/what-is-headless-mode/), custom routing will not work.
</comp.Notice>

> Custom routing allows you to control which content is delivered to your visitors when they enter a URL in their internet browser. You can create custom routing by creating a `routing.json` file in your `.cloudcannon` folder. When a visitor requests your webpage, the CloudCannon will also load the `routing.json` file to extend the default routing behavior.

Reviewer: ```suggestion
Custom routing allows you to control which content is delivered to your visitors when they enter a URL in their internet browser. You can create custom routing by creating a `routing.json` file in your `.cloudcannon` folder. When a visitor requests your webpage, CloudCannon will also load the `routing.json` file to extend the default routing behavior.
```

3. Click the *Save* button in the top right of the editing interface. CloudCannon will open the *Review changes* modal.
4. Edit the file path to "/.cloudcannon/routing.json" and click the *Save selected* button in the bottom right of the modal.

CloudCannon will save a blank `routing.json` file in your `.cloudcannon` folder.

If you would prefer, you can also create a `routing.json` file using your local development environment.

For a complete list of options available in the `routing.json` file, please read our [routing reference](/documentation/articles/routing-reference/) documentation.

## Custom Routes

When CloudCannon loads the `routing.json` file each time a visitor requests a URL, the first matching routing rule can control what content is delivered to the visitor by redirecting them to a different URL or serving a different content without changing the URL (i.e., a proxy). You can specify the type of custom route using an [HTTP status code](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes).

CloudCannon supports the following redirect codes:

> * **301** — Indicates the requested URL has been changed permanently and specifies the new URL in the response.
> * **302** — Indicates the requested URL has been changed temporarily, and the file server should continue to use the original URL in future requests.
> * **307** — Indicates the requested URL has been changed temporarily and the file server should continue to use the original URL in future requests. This is similar to a 302 code however it does not allow the HTTP method to change.
> * **308** — Indicates the requested URL has been changed permanently and the file server should continue to use the original URL in future requests. This is similar to a 301 code however it does not allow the HTTP method to change.

Reviewer: ```suggestion
* **301** — Indicates the requested URL has been changed permanently.
* **302** — Indicates the requested URL has been changed temporarily, and the visitor's browser should continue to use the original URL in future requests.
* **307** — Indicates the requested URL has been changed temporarily and the visitor's browser should continue to use the original URL in future requests. This is similar to a 302 code however it does not allow the HTTP method to change.
* **308** — Indicates the requested URL has been changed permanently. This is similar to a 301 code however it does not allow the HTTP method to change.
```

When configuring custom routing, most users will only need the 307, 308, and 404 status codes.

We recommend redirecting your old page URLs with 307 or 308 redirects whenever you change the URL of a page or the architecture of your Site in a way that affects the URL structure. Redirecting old URLs will:

* Preserve your SEO scores if you want to rank highly in search engine results.
* Maintain external links to your website.
* Prevent visitors from seeing incorrect or missing content when they attempt to load any of your old URLs.

We also recommend [configuring a 404 page](/documentation/articles/configure-a-404-page/) and routing all unrecognized subpaths to that page.

<comp.Notice info_type="info">
  CloudCannon reads your routes in order, and will use the earliest matching rule for a given `from` URL. Duplicate rules are ignored.
</comp.Notice>

<comp.Notice info_type="info">
  You can redirect to, but not from a URL fragment (e.g. example.com/blog/#heading). The URL fragment is not passed to the server so, it cannot be processed by the redirect.
</comp.Notice>

<comp.Notice info_type="info">

>   You can use routing to load a single HTML file on multiple URL routes and use JavaScript to load different content.

Reviewer: ```suggestion
  You can use routing to load a single HTML file on multiple URL routes and use JavaScript to load different content. (For example, as a single-page app, or SPA).
```

Reviewer: (to catch someone searching our docs for SPA support)

</comp.Notice>

> You must configure three keys for a custom redirect: the `from` URL, the `to` URL, and the `status` code. Proxies do not require the `to` key.

Reviewer: ```suggestion
You must configure three keys for a custom redirect: the `from` URL, the `to` URL, and the `status` code.
```

yes they do

* **From** — Specifies which URL should trigger this redirect rule.

> * **To** — Specifies which URL the visitor should be redirected to.

Reviewer: ```suggestion
* **To** — Specifies which URL the visitor should be redirected or proxied to.
```

Here are a few examples:

<comp.CodeBlock language="json" source=".cloudcannon/routing.json">
```

{
  "routes": [
    {
      "from": "/cloudcannon/(.*)",
      "to": "https://cloudcannon.com/$1",
      "status": 302
    },
    {
      "from": "/documentation/articles/choosing-your-ssg/",
      "to": "/documentation/articles/select-your-ssg/",
      "status": 301
    },
    {
      "from": "/app/(.*)",
      "to": "/app/",
      "status": 200
    }
  ]
}

```
</comp.CodeBlock>

For more a complete list of configuration keys available for custom routing, please read our [routing reference documentation](/documentation/articles/routing-reference/).

### Generating redirects

If your Site requires it, you can generate your routing file at build time.

For example, if you only want to redirect URLs on your production Site, you could add the following Liquid templating:

<comp.CodeBlock language="JSON" source="routing.json">
``````````

{
  routes: [
    {% if environment == 'production' %}
      {from: '/example-1',
      to: '/example-2'}
    {% endif %}
  ]
}

``````````
</comp.CodeBlock>

> You can create a build process to generate your routing file with the correct content, depending on the environment. In this case, the build process should generate the file in  `_cloudcannon/routing.json` . CloudCannon will use this file instead of  `.cloudcannon/routing.json` from your source.

Reviewer: ```suggestion
You can create a build process to generate your routing file with the correct content, depending on the environment. In this case, the build process should generate the file to `_cloudcannon/routing.json` in your site output directory. CloudCannon will use this file instead of  `.cloudcannon/routing.json` from your source.
```
